### PR TITLE
Introduce raw repo

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: repository/src/raw.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        -> Result<(), Error>;
+
+    /// Deletes the branch.
+    async fn delete_branch(&mut self, branch: &Branch) -> Result<(), Error>;
+
+    // -------------------
+    // Tag-related methods
+    // -------------------
+
+    /// Returns the list of branches.
```
Comment (JeongHunP) [inline repository/src/raw.rs:70]:
```suggestion
    /// Returns the list of tags.
```

File: repository/src/raw.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    /// Returns the diff of the given commit.
+    async fn show_commit(&self, commit_hash: &CommitHash) -> Result<String, Error>;
+
+    /// Lists the ancestor commits of the given commit (The first element is the direct parent).
+    ///
+    /// It fails if there is a merge commit.
+    async fn list_ancestors(
+        &self,
+        commit_hash: &CommitHash,
+        max: Option<usize>,
```
Comment (JeongHunP) [inline repository/src/raw.rs:147]: Is "max" used to restrict the number of ancestor commits?

File: repository/src/raw.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    async fn add_remote(&mut self, remote_name: &str, remote_url: &str) -> Result<(), Error>;
+
+    /// Removes a remote repository.
+    async fn remove_remote(&mut self, remote_name: &str) -> Result<(), Error>;
+
+    /// Fetches the remote repository. Same as `git fetch --all -j <LARGE NUMBER>`.
+    async fn fetch_all(&mut self) -> Result<(), Error>;
+
+    /// Lists all the remote repositories.
+    async fn list_remotes(&self) -> Result<Vec<(String, String)>, Error>;
```
Comment (JeongHunP) [inline repository/src/raw.rs:183]: Does Vec<(String, String)> match to (remote_name, remote_url)?

File: repository/src/raw.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+impl From<git2::Error> for Error {
+    fn from(e: git2::Error) -> Self {
+        Error::Git2Error(e)
+    }
+}
+
+/// A raw handle for the local repository.
+///
+/// It automatically locks the repository once created.
```
[the remark this answers is not in the document]
Comment (junha1) [reply repository/src/raw.rs:24]: https://en.wikipedia.org/wiki/File_locking

[the remark this answers is not in the document]
Comment (junha1) [reply repository/src/raw.rs:24]: you don't have to consider it when you implement the methods. I will introduce another PR that show how the locks work.

File: repository/src/raw.rs (hunk 2 above)
Comment (junha1) [reply repository/src/raw.rs:147]: comment added.

**Review by JeongHunP — APPROVED** [review]

**Pull request merged**